solutions/gold/usaco-696.mdx

```diff
@@ -306,10 +307,122 @@ int main() {
 </CPPSection>
 </LanguageSection>
 
-<IncompleteSection>
+## Solution 3: HLD
 
-Add Centroid Decomp/HLD Solution
+**Time Complexity:** $\mathcal{O}(N\log ^2N)$
+
+Sort the nodes from largest to smallest value, do path updates from each node to the root.
+
+<LanguageSection>
+<CPPSection>
+
+```cpp
+#include <bits/stdc++.h>
+using namespace std;
+
+template <class T> class SumSegmentTree {
+  private:
+	T DEFAULT = 0;
+
+	vector<T> segtree;
+	int len;
+
+  public:
+	SumSegmentTree(int len) : len(len), segtree(len * 2, DEFAULT) {}
+
+	void set(int ind, T val) {
+		ind += len;
+		segtree[ind] = val;
+		for (; ind > 1; ind /= 2) {
+			segtree[ind / 2] = segtree[ind] + segtree[ind ^ 1];
+		}
+	}
+
+	T range_sum(int start, int end) {
+		T sum = DEFAULT;
+		for (start += len, end += len; start < end; start /= 2, end /= 2) {
+			if (start % 2 == 1) { sum += segtree[start++]; }
+			if (end % 2 == 1) { sum += segtree[--end]; }
+		}
+		return sum;
+	}
+};
+
+void dfs_hld(vector<vector<int>> &edges, vector<vector<int>> &heavy_paths,
+             vector<int> &depth, vector<int> &sizes, int start) {
+	int heavy_child = -1;
+	++sizes[start];
+	for (int i : edges[start]) {
+		depth[i] = depth[start] + 1;
+		dfs_hld(edges, heavy_paths, depth, sizes, i);
+		sizes[start] += sizes[i];
+		if (heavy_child == -1 || sizes[heavy_child] < sizes[i]) heavy_child = i;
+	}
+	if (heavy_child != -1) {
+		swap(heavy_paths[start], heavy_paths[heavy_child]);
+		heavy_paths[start].push_back(heavy_child);
+	}
+}
 
-Sort nodes from largest to smallest value, do path updates from itself the root.
+int main() {
+	freopen("promote.in", "r", stdin);
+	freopen("promote.out", "w", stdout);
+	int N, current_value;
+	cin >> N;
+	vector<pair<int, int>> cow_proficiency(N);
+	for (int i = 0; i < N; ++i) {
+		cin >> cow_proficiency[i].first;
+		cow_proficiency[i].second = i;
+	}
+	vector edges(N, vector<int>()), heavy_paths(N, vector<int>());
+
+	vector<int> sizes(N, 0), heavy_index(N, -1), path_index(N, -1), depth(N, 0),
+	    parents(N, -1), ans(N, 0), cur_ans(N, 0);
 
-</IncompleteSection>
+	for (int i = 1; i < N; ++i) {
+		cin >> parents[i];
+		edges[--parents[i]].push_back(i);
+	}
+
+	dfs_hld(edges, heavy_paths, depth, sizes, 0);
+	vector segtrees(N, SumSegmentTree<int>(0));
+	for (int i = 0; i < N; ++i) {
+		reverse(heavy_paths[i].begin(), heavy_paths[i].end());
+		for (int j = 0; j < (int)heavy_paths[i].size(); ++j) {
+			heavy_index[heavy_paths[i][j]] = i;
+			path_index[heavy_paths[i][j]] = j;
+		}
+		segtrees[i] = SumSegmentTree<int>((int)heavy_paths[i].size());
+	}
+
+	sort(cow_proficiency.begin(), cow_proficiency.end(), greater<pair<int, int>>());
+	for (auto &p : cow_proficiency) {
```

Comment:
consider using a structured binding

e.g. `for (auto &[thing_a, thing_b] : cow_proficiency)` or something